### PR TITLE
kubeadm: fix the bug that configurable KubernetesVersion not respected during kubeadm join

### LINKED ISSUE #2713: kubeadm: configurable KubernetesVersion not respected during kubeadm join
### What happened?

We are using the `kubeadmConfigSpec.kubernetesVersion` configuration interface in cluster-api to enable kubeadm usage with CI builds of Kubernetes. For example:

https://github.com/kubernetes-sigs/cluster-api-provider-azure/blob/9e793b42f79b65cbc830e711dce63a2c1e7c82c0/templates/test/ci/cluster-template-prow-ci-version-windows-containerd-2022.yaml#L90

We're observing that this works file for the first kubeadm-built control plane node (e.g., `kubeadm init`) but is not working for additional control plane nodes (e.g., `kubeadm join`). This is what we're seeing in cloud-init log output for nodes that fail to kubeadm joine:

```
[2022-06-17 14:38:39] error execution phase preflight: [preflight] Some fatal errors occurred:
[2022-06-17 14:38:39]   [ERROR ImagePull]: failed to pull image registry.k8s.io/kube-apiserver:v1.25.0-alpha.1.29_19c8a2127177b4: output: time="2022-06-17T14:38:34Z" level=fatal msg="pulling image: rpc error: code = NotFound desc = failed to pull and unpack image \"registry.k8s.io/kube-apiserver:v1.25.0-alpha.1.29_19c8a2127177b4\": failed to resolve reference \"registry.k8s.io/kube-apiserver:v1.25.0-alpha.1.29_19c8a2127177b4\": registry.k8s.io/kube-apiserver:v1.25.0-alpha.1.29_19c8a2127177b4: not found"
[2022-06-17 14:38:39] , error: exit status 1
[2022-06-17 14:38:39]   [ERROR ImagePull]: failed to pull image registry.k8s.io/kube-controller-manager:v1.25.0-alpha.1.29_19c8a2127177b4: output: time="2022-06-17T14:38:36Z" level=fatal msg="pulling image: rpc error: code = NotFound desc = failed to pull and unpack image \"registry.k8s.io/kube-controller-manager:v1.25.0-alpha.1.29_19c8a2127177b4\": failed to resolve reference \"registry.k8s.io/kube-controller-manager:v1.25.0-alpha.1.29_19c8a2127177b4\": registry.k8s.io/kube-controller-manager:v1.25.0-alpha.1.29_19c8a2127177b4: not found"
[2022-06-17 14:38:39] , error: exit status 1
[2022-06-17 14:38:39]   [ERROR ImagePull]: failed to pull image registry.k8s.io/kube-scheduler:v1.25.0-alpha.1.29_19c8a2127177b4: output: time="2022-06-17T14:38:37Z" level=fatal msg="pulling image: rpc error: code = NotFound desc = failed to pull and unpack image \"registry.k8s.io/kube-scheduler:v1.25.0-alpha.1.29_19c8a2127177b4\": failed to resolve reference \"registry.k8s.io/kube-scheduler:v1.25.0-alpha.1.29_19c8a2127177b4\": registry.k8s.io/kube-scheduler:v1.25.0-alpha.1.29_19c8a2127177b4: not found"
[2022-06-17 14:38:39] , error: exit status 1
[2022-06-17 14:38:39]   [ERROR ImagePull]: failed to pull image registry.k8s.io/kube-proxy:v1.25.0-alpha.1.29_19c8a2127177b4: output: time="2022-06-17T14:38:39Z" level=fatal msg="pulling image: rpc error: code = NotFound desc = failed to pull and unpack image \"registry.k8s.io/kube-proxy:v1.25.0-alpha.1.29_19c8a2127177b4\": failed to resolve reference \"registry.k8s.io/kube-proxy:v1.25.0-alpha.1.29_19c8a2127177b4\": registry.k8s.io/kube-proxy:v1.25.0-alpha.1.29_19c8a2127177b4: not found"
[2022-06-17 14:38:39] , error: exit status 1
```

### What did you expect to happen?

`kubeadm init` and `kubeadm join` should have similar outcomes based on `kubeadmConfigSpec.kubernetesVersion` configuration.

### How can we reproduce it (as minimally and precisely as possible)?

Build any cluster-api cluster and configure the `KubeadmControlPlane`'s `kubeadmConfigSpec.kubernetesVersion` to a CI build, for example `ci/v1.25.0-alpha.0.734_ba502ee555924a`, and set the `replicas` count to 3.

The first control plane node will come online and become Ready, but the remaining control plane nodes will not.

### Anything else we need to know?

_No response_

### Kubernetes version

<details>
Any Kubernetes CI version, e.g., the version returned from the https://storage.googleapis.com/k8s-release-dev/ci/latest-1.25.txt URL.
</details>

### Cloud provider

<details>
verified on cluster-api using capz and capa
</details>

### OS version

<details>

```console
# On Linux:
$ cat /etc/os-release
# paste output here
$ uname -a
# paste output here

# On Windows:
C:\> wmic os get Caption, Version, BuildNumber, OSArchitecture
# paste output here
```

</details>

### Install tools

<details>

</details>

### Container runtime (CRI) and version (if applicable)

<details>

</details>

### Related plugins (CNI, CSI, ...) and versions (if applicable)

<details>

</details>

### PR DESCRIPTION
**Pull request by SataQiu**
#### What type of PR is this?
/kind bug

#### What this PR does / why we need it:
kubeadm: fix the bug that configurable KubernetesVersion not respected during kubeadm join

#### Which issue(s) this PR fixes:

Fixes kubernetes/kubeadm#2713

#### Does this PR introduce a user-facing change?

```release-note
kubeadm: fix the bug that configurable KubernetesVersion not respected during kubeadm join
```

#### Additional documentation e.g., KEPs (Kubernetes Enhancement Proposals), usage docs, etc.:

```docs

```

### CODE REVIEW

File: cmd/kubeadm/app/util/version.go [hunk 1]
```diff
@@ -144,6 +144,14 @@ func KubernetesIsCIVersion(version string) bool {
 	if len(subs) == 1 && len(subs[0]) == 4 && strings.HasPrefix(subs[0][2], "ci") {
 		return true
 	}
+	ver, err := versionutil.ParseSemantic(version)
+	if err != nil {
+		klog.Warningf("could not parse semantic version %s: %v", version, err)
+		return false
+	}
+	if len(ver.PreRelease()) != 0 {
```
Comment (pacoxu) [inline cmd/kubeadm/app/util/version.go:152]:
```suggestion
	if len(ver.PreRelease()) != 0 && ver != constants.DefaultKubernetesPlaceholderVersion {
```
`DefaultKubernetesPlaceholderVersion` is now `defaultKubernetesPlaceholderVersion`

v1.0.0-placeholder-version is a place holder, not ci version.

Comment (pacoxu) [reply cmd/kubeadm/app/util/version.go:152]:
```
diff --git a/cmd/kubeadm/app/constants/constants.go b/cmd/kubeadm/app/constants/constants.go
index 124295d0512..87342950fc1 100644
--- a/cmd/kubeadm/app/constants/constants.go
+++ b/cmd/kubeadm/app/constants/constants.go
@@ -508,9 +508,9 @@ var (
                Jitter:   0.1,
        }
 
-       // defaultKubernetesPlaceholderVersion is a placeholder version in case the component-base
+       // DefaultKubernetesPlaceholderVersion is a placeholder version in case the component-base
        // version was not populated during build.
-       defaultKubernetesPlaceholderVersion = version.MustParseSemantic("v1.0.0-placeholder-version")
+       DefaultKubernetesPlaceholderVersion = version.MustParseSemantic("v1.0.0-placeholder-version")
 )
 
 // getSkewedKubernetesVersion returns the current MAJOR.(MINOR+n).0 Kubernetes version with a skew of 'n'
@@ -528,7 +528,7 @@ func getSkewedKubernetesVersionImpl(versionInfo *apimachineryversion.Info, n int
        // More changes would be required if the kubelet version one day decouples from that of Kubernetes.
        var ver *version.Version
        if len(versionInfo.Major) == 0 {
-               return defaultKubernetesPlaceholderVersion
+               return DefaultKubernetesPlaceholderVersion
        }
        ver = version.MustParseSemantic(versionInfo.GitVersion)
        // Append the MINOR version skew.
diff --git a/cmd/kubeadm/app/constants/constants_test.go b/cmd/kubeadm/app/constants/constants_test.go
index 97046c592a3..afa68689911 100644
--- a/cmd/kubeadm/app/constants/constants_test.go
+++ b/cmd/kubeadm/app/constants/constants_test.go
@@ -244,7 +244,7 @@ func TestGetSkewedKubernetesVersionImpl(t *testing.T) {
                {
                        name:           "invalid versionInfo; placeholder version is returned",
                        versionInfo:    &apimachineryversion.Info{},
-                       expectedResult: defaultKubernetesPlaceholderVersion,
+                       expectedResult: DefaultKubernetesPlaceholderVersion,
                },
                {
                        name:           "valid skew of -1",
diff --git a/cmd/kubeadm/app/util/version.go b/cmd/kubeadm/app/util/version.go
index 744c4a810fa..a50d5d8f43e 100644
--- a/cmd/kubeadm/app/util/version.go
+++ b/cmd/kubeadm/app/util/version.go
@@ -149,7 +149,7 @@ func KubernetesIsCIVersion(version string) bool {
                klog.Warningf("could not parse semantic version %s: %v", version, err)
                return false
        }
-       if len(ver.PreRelease()) != 0 {
+       if len(ver.PreRelease()) != 0 && ver.String() != constants.DefaultKubernetesPlaceholderVersion.String() {
                return true
        }
        return false
```

**Comment by neolit123** [comment]
added some comments here:
https://github.com/kubernetes/kubeadm/issues/2713#issuecomment-1167382278

File: cmd/kubeadm/app/util/config/common.go [hunk 2]
```diff
@@ -100,6 +100,7 @@ func NormalizeKubernetesVersion(cfg *kubeadmapi.ClusterConfiguration) error {
 	if err != nil {
 		return err
 	}
+	cfg.UnNormalizedKubernetesVersion = cfg.KubernetesVersion
```
[the remark this answers is not in the document]
Comment (SataQiu) [reply cmd/kubeadm/app/util/config/common.go:103]: Yes, you are right. It seems that we should NOT save `ci/latest` directly in this scenario.

**Review by neolit123 — COMMENTED** [review]
/lgtm
This seems better to me. Thanks.

/hold
@pacoxu does this seem right to you?

**Comment by pacoxu** [comment]
/lgtm
/unhold

There are already some unit tests in https://github.com/pacoxu/kubernetes/blob/8ae6a2048573041766bc103fba106268e78e8460/cmd/kubeadm/app/util/version_test.go  that can validate this.

**Comment by jackfrancis** [comment]
@SataQiu @neolit123 @pacoxu I'm seeing some failures in conformance tests against latest-1.24, perhaps since this was cherry-picked and a new build pushed with these changes?

```
[2022-07-07 11:26:24] kubectl version:  Client Version: v1.24.3-rc.0.27+7ca32977924bc1 Kustomize Version: v4.5.4
[2022-07-07 11:26:24] kubelet version:  Kubernetes v1.24.3-rc.0.27+7ca32977924bc1
[2022-07-07 11:26:24] *************************************************
[2022-07-07 11:26:24] [preflight] Running pre-flight checks
[2022-07-07 11:26:25] [preflight] Reading configuration from the cluster...
[2022-07-07 11:26:25] [preflight] FYI: You can look at this config file with 'kubectl -n kube-system get cm kubeadm-config -o yaml'
[2022-07-07 11:26:25] error execution phase preflight: unable to fetch the kubeadm-config ConfigMap: failed to get component configs: could not parse "ci/v1.24.3-rc.0.27+7ca32977924bc1" as version
```

Does the above look related to these changes?

This is on a worker node btw, not a control plane node. Control plane node looks good:

```
$ k get nodes -o wide
NAME                                   STATUS   ROLES           AGE   VERSION                          INTERNAL-IP   EXTERNAL-IP   OS-IMAGE             KERNEL-VERSION     CONTAINER-RUNTIME
capz-conf-0fmzec-control-plane-sr69v   Ready    control-plane   11m   v1.24.3-rc.0.27+7ca32977924bc1   10.0.0.4      <none>        Ubuntu 18.04.6 LTS   5.4.0-1085-azure   containerd://1.6.2
```

**Comment by neolit123** [comment]
@SataQiu the version parser should work but perhaps something is missing and maybe we don't have a unit test.

**Comment by jackfrancis** [comment]
Repro cluster configmap below, certainly seems the result of these changes:

```
s$ k get configmaps kubeadm-config -n kube-system -o yaml
apiVersion: v1
data:
  ClusterConfiguration: |
    apiServer:
      extraArgs:
        authorization-mode: Node,RBAC
        cloud-config: /etc/kubernetes/azure.json
        cloud-provider: azure
        feature-gates: ""
      extraVolumes:
      - hostPath: /etc/kubernetes/azure.json
        mountPath: /etc/kubernetes/azure.json
        name: cloud-config
        readOnly: true
      timeoutForControlPlane: 20m0s
    apiVersion: kubeadm.k8s.io/v1beta3
    certificatesDir: /etc/kubernetes/pki
    clusterName: capz-conf-3ikdkt
    controlPlaneEndpoint: capz-conf-3ikdkt-8249ef04.westus2.cloudapp.azure.com:6443
    controllerManager:
      extraArgs:
        allocate-node-cidrs: "false"
        cloud-config: /etc/kubernetes/azure.json
        cloud-provider: azure
        cluster-name: capz-conf-3ikdkt
        feature-gates: HPAContainerMetrics=true
        v: "4"
      extraVolumes:
      - hostPath: /etc/kubernetes/azure.json
        mountPath: /etc/kubernetes/azure.json
        name: cloud-config
        readOnly: true
    dns: {}
    etcd:
      local:
        dataDir: /var/lib/etcddisk/etcd
        extraArgs:
          quota-backend-bytes: "8589934592"
    imageRepository: k8s.gcr.io
    kind: ClusterConfiguration
    kubernetesVersion: ci/v1.24.3-rc.0.27+7ca32977924bc1
    networking:
      dnsDomain: cluster.local
      podSubnet: 192.168.0.0/16
      serviceSubnet: 10.96.0.0/12
    scheduler: {}
kind: ConfigMap
metadata:
  creationTimestamp: "2022-07-07T11:53:24Z"
  name: kubeadm-config
  namespace: kube-system
  resourceVersion: "235"
  uid: 172efec9-42aa-4d13-b2a7-a3e3fadae3b6
```

**Comment by jackfrancis** [comment]
Confirmed the obvious:

```
$ git diff staging/src/k8s.io/apimachinery/pkg/util/version/version_test.go
diff --git a/staging/src/k8s.io/apimachinery/pkg/util/version/version_test.go b/staging/src/k8s.io/apimachinery/pkg/util/version/version_test.go
index aa0675f7f12..1ea5a75de27 100644
--- a/staging/src/k8s.io/apimachinery/pkg/util/version/version_test.go
+++ b/staging/src/k8s.io/apimachinery/pkg/util/version/version_test.go
@@ -96,6 +96,7 @@ func TestSemanticVersions(t *testing.T) {
                {version: "2.1.0"},
                {version: "2.1.1"},
                {version: "42.0.0"},
+               {version: "ci/v1.24.3-rc.0.27+7ca32977924bc1"},
 
                // We also allow whitespace and "v" prefix
                {version: "   42.0.0", unparsed: "42.0.0", equalsPrev: true},

go test -timeout 10m -tags e2e -run ^TestSemanticVersions$ k8s.io/apimachinery/pkg/util/version

--- FAIL: TestSemanticVersions (0.00s)
    /Users/jackfrancis/go/src/github.com/kubernetes/kubernetes/staging/src/k8s.io/apimachinery/pkg/util/version/version_test.go:115: unexpected parse error: could not parse "ci/v1.24.3-rc.0.27+7ca32977924bc1" as version
```

**Comment by neolit123** [comment]
The e2e tests pin a k8s version that is without ci prefixes. I guess that is mostly because we bake our own images from tars and we treat it as a regular k8s version, not ci.

It should be possible to test this scenario with unit tests. Or modify our e2e tests somehow (probably harder to do).

**Comment by SataQiu** [comment]
Hi @jackfrancis , I guess you probably didn't update the `kubeadm` binary for the worker node.
I ran into this problem if using the old kubeadm. 
This problem no longer occurs when I use the latest version.

int control-plane node:
```bash
root@kind-control-plane:/# ./kubeadm version
kubeadm version: &version.Info{Major:"1", Minor:"25+", GitVersion:"v1.25.0-alpha.0.1261+d3092cd296f82d-dirty", GitCommit:"d3092cd296f82d24236f57f5928cd4755a080d5c", GitTreeState:"dirty", BuildDate:"2022-07-08T06:45:45Z", GoVersion:"go1.18.1", Compiler:"gc", Platform:"linux/arm64"}
root@kind-control-plane:/# ./kubeadm init --kubernetes-version=ci/latest-1.24
[init] Using Kubernetes version: v1.24.3-rc.0.29+99b7713ba36e8f
[preflight] Running pre-flight checks
	[WARNING Swap]: swap is enabled; production deployments should disable swap unless testing the NodeSwap feature gate of the kubelet
	[WARNING SystemVerification]: missing optional cgroups: blkio
[preflight] Pulling images required for setting up a Kubernetes cluster
[preflight] This might take a minute or two, depending on the speed of your internet connection
[preflight] You can also perform this action in beforehand using 'kubeadm config images pull'
[certs] Using certificateDir folder "/etc/kubernetes/pki"
[certs] Generating "ca" certificate and key
[certs] Generating "apiserver" certificate and key
[certs] apiserver serving cert is signed for DNS names [kind-control-plane kubernetes kubernetes.default kubernetes.default.svc kubernetes.default.svc.cluster.local] and IPs [10.96.0.1 172.18.0.3]
[certs] Generating "apiserver-kubelet-client" certificate and key
[certs] Generating "front-proxy-ca" certificate and key
[certs] Generating "front-proxy-client" certificate and key
[certs] Generating "etcd/ca" certificate and key
[certs] Generating "etcd/server" certificate and key
[certs] etcd/server serving cert is signed for DNS names [kind-control-plane localhost] and IPs [172.18.0.3 127.0.0.1 ::1]
[certs] Generating "etcd/peer" certificate and key
[certs] etcd/peer serving cert is signed for DNS names [kind-control-plane localhost] and IPs [172.18.0.3 127.0.0.1 ::1]
[certs] Generating "etcd/healthcheck-client" certificate and key
[certs] Generating "apiserver-etcd-client" certificate and key
[certs] Generating "sa" key and public key
[kubeconfig] Using kubeconfig folder "/etc/kubernetes"
[kubeconfig] Writing "admin.conf" kubeconfig file
[kubeconfig] Writing "kubelet.conf" kubeconfig file
[kubeconfig] Writing "controller-manager.conf" kubeconfig file
[kubeconfig] Writing "scheduler.conf" kubeconfig file
[kubelet-start] Writing kubelet environment file with flags to file "/var/lib/kubelet/kubeadm-flags.env"
[kubelet-start] Writing kubelet configuration to file "/var/lib/kubelet/config.yaml"
[kubelet-start] Starting the kubelet
[control-plane] Using manifest folder "/etc/kubernetes/manifests"
[control-plane] Creating static Pod manifest for "kube-apiserver"
[control-plane] Creating static Pod manifest for "kube-controller-manager"
[control-plane] Creating static Pod manifest for "kube-scheduler"
[etcd] Creating static Pod manifest for local etcd in "/etc/kubernetes/manifests"
[wait-control-plane] Waiting for the kubelet to boot up the control plane as static Pods from directory "/etc/kubernetes/manifests". This can take up to 4m0s
[apiclient] All control plane components are healthy after 5.503461 seconds
[upload-config] Storing the configuration used in ConfigMap "kubeadm-config" in the "kube-system" Namespace
[kubelet] Creating a ConfigMap "kubelet-config" in namespace kube-system with the configuration for the kubelets in the cluster
[upload-certs] Skipping phase. Please see --upload-certs
[mark-control-plane] Marking the node kind-control-plane as control-plane by adding the labels: [node-role.kubernetes.io/control-plane node.kubernetes.io/exclude-from-external-load-balancers]
[mark-control-plane] Marking the node kind-control-plane as control-plane by adding the taints [node-role.kubernetes.io/control-plane:NoSchedule]
[bootstrap-token] Using token: ri7v1v.ctuj2k2l9pt17id4
[bootstrap-token] Configuring bootstrap tokens, cluster-info ConfigMap, RBAC Roles
[bootstrap-token] Configured RBAC rules to allow Node Bootstrap tokens to get nodes
[bootstrap-token] Configured RBAC rules to allow Node Bootstrap tokens to post CSRs in order for nodes to get long term certificate credentials
[bootstrap-token] Configured RBAC rules to allow the csrapprover controller automatically approve CSRs from a Node Bootstrap Token
[bootstrap-token] Configured RBAC rules to allow certificate rotation for all node client certificates in the cluster
[bootstrap-token] Creating the "cluster-info" ConfigMap in the "kube-public" namespace
[kubelet-finalize] Updating "/etc/kubernetes/kubelet.conf" to point to a rotatable kubelet client certificate and key
[addons] Applied essential addon: CoreDNS
[addons] Applied essential addon: kube-proxy

Your Kubernetes control-plane has initialized successfully!

To start using your cluster, you need to run the following as a regular user:

  mkdir -p $HOME/.kube
  sudo cp -i /etc/kubernetes/admin.conf $HOME/.kube/config
  sudo chown $(id -u):$(id -g) $HOME/.kube/config

Alternatively, if you are the root user, you can run:

  export KUBECONFIG=/etc/kubernetes/admin.conf

You should now deploy a pod network to the cluster.
Run "kubectl apply -f [podnetwork].yaml" with one of the options listed at:
  https://kubernetes.io/docs/concepts/cluster-administration/addons/

Then you can join any number of worker nodes by running the following on each as root:

kubeadm join 172.18.0.3:6443 --token ri7v1v.ctuj2k2l9pt17id4 \
	--discovery-token-ca-cert-hash sha256:3eb72a6cfa9616f8c1b8f9b061a91255e98add0870c966b84d7932e41a8a905f 
```

check kubeadm-config:
```bash
root@kind-control-plane:/# kubectl get cm -n kube-system kubeadm-config -oyaml
apiVersion: v1
data:
  ClusterConfiguration: |
    apiServer:
      extraArgs:
        authorization-mode: Node,RBAC
      timeoutForControlPlane: 4m0s
    apiVersion: kubeadm.k8s.io/v1beta3
    certificatesDir: /etc/kubernetes/pki
    clusterName: kubernetes
    controllerManager: {}
    dns: {}
    etcd:
      local:
        dataDir: /var/lib/etcd
    imageRepository: registry.k8s.io
    kind: ClusterConfiguration
    kubernetesVersion: ci/v1.24.3-rc.0.29+99b7713ba36e8f
    networking:
      dnsDomain: cluster.local
      serviceSubnet: 10.96.0.0/12
    scheduler: {}
kind: ConfigMap
metadata:
  creationTimestamp: "2022-07-08T08:01:43Z"
  name: kubeadm-config
  namespace: kube-system
  resourceVersion: "204"
  uid: 6022a1cf-7f75-4a2c-99f5-607b0c75fbae
```

join worker node use old kubeadm:
```bash
root@kind-worker:/# kubeadm version
kubeadm version: &version.Info{Major:"1", Minor:"24", GitVersion:"v1.24.0", GitCommit:"4ce5a8954017644c5420bae81d72b09b735c21f0", GitTreeState:"clean", BuildDate:"2022-05-19T15:42:59Z", GoVersion:"go1.18.1", Compiler:"gc", Platform:"linux/arm64"}
root@kind-worker:/# kubeadm join 172.18.0.3:6443 --token ri7v1v.ctuj2k2l9pt17id4 \
        --discovery-token-ca-cert-hash sha256:3eb72a6cfa9616f8c1b8f9b061a91255e98add0870c966b84d7932e41a8a905f 
[preflight] Running pre-flight checks
	[WARNING Swap]: swap is enabled; production deployments should disable swap unless testing the NodeSwap feature gate of the kubelet
	[WARNING SystemVerification]: missing optional cgroups: blkio
[preflight] Reading configuration from the cluster...
[preflight] FYI: You can look at this config file with 'kubectl -n kube-system get cm kubeadm-config -o yaml'
error execution phase preflight: unable to fetch the kubeadm-config ConfigMap: failed to get component configs: could not parse "ci/v1.24.3-rc.0.29+99b7713ba36e8f" as version
To see the stack trace of this error execute with --v=5 or higher
```

join worker using new kubeadm:
```bash
root@kind-worker:/# ./kubeadm version
kubeadm version: &version.Info{Major:"1", Minor:"25+", GitVersion:"v1.25.0-alpha.0.1261+d3092cd296f82d-dirty", GitCommit:"d3092cd296f82d24236f57f5928cd4755a080d5c", GitTreeState:"dirty", BuildDate:"2022-07-08T06:45:45Z", GoVersion:"go1.18.1", Compiler:"gc", Platform:"linux/arm64"}
root@kind-worker:/# ./kubeadm join 172.18.0.3:6443 --token ri7v1v.ctuj2k2l9pt17id4      --discovery-token-ca-cert-hash sha256:3eb72a6cfa9616f8c1b8f9b061a91255e98add0870c966b84d7932e41a8a905f 
[preflight] Running pre-flight checks
	[WARNING Swap]: swap is enabled; production deployments should disable swap unless testing the NodeSwap feature gate of the kubelet
	[WARNING SystemVerification]: missing optional cgroups: blkio
[preflight] Reading configuration from the cluster...
[preflight] FYI: You can look at this config file with 'kubectl -n kube-system get cm kubeadm-config -o yaml'
[kubelet-start] Writing kubelet configuration to file "/var/lib/kubelet/config.yaml"
[kubelet-start] Writing kubelet environment file with flags to file "/var/lib/kubelet/kubeadm-flags.env"
[kubelet-start] Starting the kubelet
[kubelet-start] Waiting for the kubelet to perform the TLS Bootstrap...

This node has joined the cluster:
* Certificate signing request was sent to apiserver and a response was received.
* The Kubelet was informed of the new secure connection details.

Run 'kubectl get nodes' on the control-plane to see this node join the cluster.

```

**Comment by jackfrancis** [comment]
Thanks @SataQiu, I will check that!

**Comment by jackfrancis** [comment]
From control plane node able to successfully init:

`[  141.179848] cloud-init[1842]: [2022-07-08 10:12:59] kubeadm version:  v1.24.3-rc.0.29+99b7713ba36e8f`

From a worker node that fails pre-flight check:

```
[  585.910150] cloud-init[1807]: [2022-07-08 10:16:02] kubeadm version:  v1.24.3-rc.0.29+99b7713ba36e8f
[  585.910654] cloud-init[1807]: [2022-07-08 10:16:02] Flag --short has been deprecated, and will be removed in the future. The --short output will become the default.
[  585.911603] cloud-init[1807]: [2022-07-08 10:16:02] kubectl version:  Client Version: v1.24.3-rc.0.29+99b7713ba36e8f Kustomize Version: v4.5.4
[  585.912095] cloud-init[1807]: [2022-07-08 10:16:02] kubelet version:  Kubernetes v1.24.3-rc.0.29+99b7713ba36e8f
[  585.912575] cloud-init[1807]: [2022-07-08 10:16:02] *************************************************
[  585.913051] cloud-init[1807]: [2022-07-08 10:16:02] [preflight] Running pre-flight checks
[  585.913816] cloud-init[1807]: [2022-07-08 10:16:06] [preflight] Reading configuration from the cluster...
[  585.914380] cloud-init[1807]: [2022-07-08 10:16:06] [preflight] FYI: You can look at this config file with 'kubectl -n kube-system get cm kubeadm-config -o yaml'
[  585.914879] cloud-init[1807]: [2022-07-08 10:16:06] error execution phase preflight: unable to fetch the kubeadm-config ConfigMap: failed to get component configs: could not parse "ci/v1.24.3-rc.0.29+99b7713ba36e8f" as version
```

tl;dr we are using a new version of kubeadm that has the change in our repro:

```
$ kubeadm version
kubeadm version: &version.Info{Major:"1", Minor:"24+", GitVersion:"v1.24.3-rc.0.29+99b7713ba36e8f", GitCommit:"99b7713ba36e8ffb82e2ca8d631c960853cbb637", GitTreeState:"clean", BuildDate:"2022-07-07T18:13:36Z", GoVersion:"go1.18.3", Compiler:"gc", Platform:"linux/amd64"}
```

**Comment by neolit123** [comment]
> From a worker node that fails pre-flight check:

a --v=5 would have helped here.

**Comment by jackfrancis** [comment]
@neolit123 I'll paste a verbosity=5 output from a failing node here

thanks @SataQiu! hopefully my `--v=5` output will further confirm that what we're seeing is the same thing your fix is addressing

**Comment by neolit123** [comment]
commented on https://github.com/kubernetes/kubernetes/pull/111021
for different k8s versions the fix will have to be different `< 1.25` VS `>= 1.25`
